File: Algo/Leetcode/063UniquePathsII.cpp
// dp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int>> dp(m, vector<int>(n));
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(obstacleGrid[i][j]==1) dp[i][j] = 0;
                else if(i==0 && j==0) dp[i][j] = 1;
                else if(i==0) dp[i][j] = dp[i][j-1] ;
                else if(j==0) dp[i][j] = dp[i-1][j] ;
                else dp[i][j] = dp[i-1][j] + dp[i][j-1] ;
            }
        }
        return dp[m-1][n-1];
    }
};

// use original input
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(obstacleGrid[i][j]==1) obstacleGrid[i][j] = 0;
                else if(i==0 && j==0) obstacleGrid[i][j] = 1;
                else if(i==0) obstacleGrid[i][j] = obstacleGrid[i][j-1] ;
                else if(j==0) obstacleGrid[i][j] = obstacleGrid[i-1][j] ;
                else obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1] ;
            }
        }
        return obstacleGrid[m-1][n-1];
    }
};

File: Algo/Leetcode/063UniquePathsII.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        n = len(obstacleGrid)
        m = len(obstacleGrid[0])
        obstacleGrid[0][0] = 1-obstacleGrid[0][0]
        for i in range(1, n):
            obstacleGrid[i][0] = obstacleGrid[i-1][0] if obstacleGrid[i][0]==0 else 0
        for j in range(1, m):
            obstacleGrid[0][j] = obstacleGrid[0][j-1] if obstacleGrid[0][j]==0 else 0

        for i in range(1, n):
            for j in range(1, m):
                if obstacleGrid[i][j] == 1:
                    obstacleGrid[i][j] = 0
                else:
                    obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1]

        return obstacleGrid[-1][-1]